### PR TITLE
[Feat] Inter-batch-parellelism

### LINKED ISSUE #8316: Add a flavor of `training_step` that allows for expressing inter-batch parallelism
### Motivation

When training a model, the most common approach is to process input batches sequentially (Lightning does this today). This makes sense, because in order to achieve synchronous/deterministic training, the processing of batch[n+1] can’t start until the optimization for batch[n] completes. However, this is only partly true, because it only applies to model weight-dependent logic. We can execute the model weight-independent logic out-of-order while still maintaining numerical equivalence. If the model weight-independent logic takes significant amount of wall time and stresses different system resources compared to the rest of the training step, parallelization can lead to increased throughput.

A trivial example of this is hiding the latency of host-to-device copy of input batches behind computational intensive operations. Without parallization, the execution looks like:
```
batch 0: [HtoD][forward][backward]
batch 1:                          [HtoD][forward][backward]
batch 2:                                                   [HtoD][forward][backward]
```
With parallelization, the latency of HtoD copy can be hidden:
```
batch 0: [HtoD][forward][backward]
batch 1:       [HtoD]             [forward][backward]
batch 2:             [HtoD]                          [forward][backward]
```
A more concrete use case for inter-batch parallelism at FB is to accelerate the model parallel technique used in DLRM (https://github.com/facebookresearch/dlrm). DLRM models consist large amount of embedding layers which are sharded across different GPUs. The forward pass for the embedding layers is roughly as follows:
1. All ranks receive global indices
2. The indices are routed to GPUs that hold the corresponding embedding weights (with alltoall collective)
3. All ranks perform local embedding lookup
4. The embedding lookup results are routed back to senders of the indices (with alltoall collective)

As we can see, step 1 and 2 are (1) model weight-independent (2) expensive (3) stresses communication resources. With inter-batch parallelism, we can hide the latency behind the previous batch’s backward pass.

### Pitch

Support a `training_step` flavor with access to the data iterator, so users can express the aforementioned algorithms.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #8316

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/dataloader_fetcher.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                yield
+                self.end_record()
+        else:
+            self.start_record()
+            yield
+            self.end_record()
+
+    @property
+    def cuda_inter_batch_parallelism(self) -> bool:
+        return self.device.type == "cuda" and self.inter_batch_parallelism
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/dataloader_fetcher.py:40]: Would be better to crash when `inter_batch_parallelism` for non GPU Accelerators.

Comment (tchaton) [reply pytorch_lightning/utilities/dataloader_fetcher.py:40]: Yes, good idea.

File: pytorch_lightning/utilities/dataloader_fetcher.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    iterator = iter(iterable)
+    while True:
+        try:
+            with profiler.profile("get_train_batch"):
+                yield next(iterator)
+        except StopIteration:
+            return
+
+
+class LightningFetcher:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/dataloader_fetcher.py:31]: n00b question: certain classes are prefixed with `Lightning` while others aren't.
Components with name:
- LightningModule, LightningDataModule, LightningLoggerBase
- without: Trainer, Callbacks, Logger, Profiler

utilities like this don't contain logic specific to the rest of the framework, so I wonder if we could call this just Prefetcher?

it can also help users feel like they're not needing to learn lightning-specific things

File: pytorch_lightning/utilities/dataloader_fetcher.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    def prefetch_iterator(self) -> Any:
+        cuda_stream = torch.cuda.Stream()
+
+        done = False
+        while not done:
+
+            for _ in range(self.num_prefetch_batches + 1):
+                if not done:
+                    with torch.cuda.stream(cuda_stream):
```
Comment (ananthsub) [inline pytorch_lightning/utilities/dataloader_fetcher.py:91]: should this validate somewhere that torch.cuda is available?

File: pytorch_lightning/utilities/dataloader_fetcher.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:       [HtoD]             [forward][backward]
+    """
+
+    def __init__(
+        self,
+        dataloader,
+        batch_to_device: Callable,
+        profiler: "pl.profiler.base.BaseProfiler",
```
Comment (ananthsub) [inline pytorch_lightning/utilities/dataloader_fetcher.py:49]: could the profiler be optional? the lightning trainer would always set this, but this becomes a handy utility for users outside the project too (it might even draw them into the project)

File: pytorch_lightning/utilities/dataloader_fetcher.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                yield next(iterator)
+        except StopIteration:
+            return
+
+
+class LightningFetcher:
+
+    """
+    This class is used to perform ``pre-fetching`` for the ``train`` dataloader
+    and apply inter batch parallelism if enabled.
```
Comment (ananthsub) [inline pytorch_lightning/utilities/dataloader_fetcher.py:35]: this could/should be used for evaluation too to still overlap the forward with host to device transfer

File: pytorch_lightning/utilities/dataloader_fetcher.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/dataloader_fetcher.py:91]: We validate it here as it is only supported on GPU devices https://github.com/PyTorchLightning/pytorch-lightning/blob/inter_batch_parallism/pytorch_lightning/trainer/trainer.py#L1344

File: pytorch_lightning/utilities/dataloader_fetcher.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/utilities/dataloader_fetcher.py:35]: Yes, added in fault tolerant training PRs :)

File: pytorch_lightning/utilities/dataloader_fetcher.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/dataloader_fetcher.py:31]: I would propose `DataFetcher`.
